src/main/clojure/clara/rules/test_rules_data.clj

```diff
@@ -0,0 +1,31 @@
+(ns clara.rules.test-rules-data
```

Comment:
When I put it in src/test/common, I got errors that it couldn't be found. I found that puzzling, maybe I just had a typo or something, will give it another try.

Reply:
If you still run into errors that you can't find a reason for with this in src/test/common please post them, I'll take a look and see if I have some insight into their cause(s).

Reply:
clojure.lang.ExceptionInfo: failed compiling file:src/test/clojurescript/clara/test_rules.cljs {:file #object[java.io.File 0x1a314f8b "src/test/clojurescript/clara/test_rules.cljs"]}
...
Caused by: clojure.lang.ExceptionInfo:  at line 97 src/test/clojurescript/clara/test_rules.cljs {:file "src/test/clojurescript/clara/test_rules.cljs", :line 97, :column 1, :tag :cljs/analysis-error}
...
Caused by: java.lang.NullPointerException

I assume this is because when `defsession` tries to eval `(clara.test-rules-data/weather-rules)` it's not found, because the common folder doesn't get AOT'ed while the main source folder is all compiled before the tests are run. So I tried moving it back into the clojurescript folder, where it started. Still doesn't work.

Reply:
I'm not convinced that `defsession`s use of `eval` within the macro is a good practice. I've had similarly difficult issues with upgrading the Clara project's cljs to use a more modern Figwheel setup. I plan to work on that some again soon.

In this particular case, it doesn't make sense to me that the `clara.test-rules-data` ns is being brought in via `:require-macros`. It isn't being used for macros. It is being used to call a fn.

I think that the namespace should be moved back to "src/test/common/clara/test_rules_data.cljc". So use a cljc file and then just do a `:require` on it from the cljs side. If there is any clj specifics in it try using read-conditionals.

Reply:
I absolutely agree, and explored multiple avenues, but could not get it to work in any other but the current form. I think there may be two separate issues. One is that for reasons unknown, even leaving everything exactly as is and putting test_rules_data.clj in common causes compile errors, which surprises me, but I assume is related to the same reason clara.rules.testfacts is not included in the test tree. The other reason I presume has to do with the `defsession` operation as you describe. The data used to create the session has to be available to the `defsession` macro at compile-time, which means it can't just be a `def` in CLJS or CLJC. I assume the use of `require-macros` on a CLJ file is a hack to get that data into scope at compile time. Even directly referencing the symbols like `the-rules` from `defsession` doesn't work. It has to be the form seen here, a call to a function which just returns the value referenced by the symbol. I assume that has something to do with how `eval` operates.

Reply:
I think the `clara.rules.testfacts` thing is unrelated. That was some older issue with how Clara was resolving the `defrecord` types, which I think is no longer an issue and those should have been moved. I'm pretty sure it wasn't an issue (at least not at the time) with cljs compilation.

I'm glad you pointed out that data for `defsession` would have to be available at compilation time and that `require-macros` was probably trying to workaround that. That definition sounds like an abuse of `require-macros` to me. I'm not sure this pattern of bringing in the common rules to `defsession` is a sustainable one. If anything, the rule data structure needs to be exposed via something that `defsession` is meant to consume, which I believe would be a collection of symbols referring to production data structures, or a symbol referring to a namespace (presumably with productions in it).
`defsession` is also odd in it's use of quotes on symbols in some cases, but not others. It sounds like that macro could use some work.

I think this way of passing the productions for testing to it needs work too though.

Reply:
For now, I wonder if the `clara.test-rules-data` could be clj or cljc and written like
```clj
(ns clara.test-rules-data
  (:require [clara.rules]
            [clara.rules.testfacts]))

(def the-rules
  [{:doc "Rule to determine whether it is indeed cold and windy."
    :name "clara.test-rules-data/is-cold-and-windy-data"
    :lhs '[{:type clara.rules.testfacts/Temperature
            :constraints [(< temperature 20)
                          (== ?t temperature)]}
           {:type clara.rules.testfacts/WindSpeed
            :constraints [(> windspeed 30)
                          (== ?w windspeed)]}]
    :rhs '(clara.rules/insert! (clara.rules.testfacts/->ColdAndWindy ?t ?w))}

   {:name "clara.test-rules-data/find-cold-and-windy-data"
    :lhs '[{:fact-binding :?fact
            :type clara.rules.testfacts/ColdAndWindy
            :constraints []}]
    :params #{}}])

(defmacro weather-rules
  "Return some weather rules"
  []
  `'~the-rules)
```

Then in an example, like `clara.test-rules` on the cljs side
Add `:require` on `[clara.test-rules-data :as trd :include-macros true]`

then use like
```clj

(defsession my-session-data (trd/weather-rules))
```

Reply:
I tried something like that and couldn't get it to work, though to be fair there were potentially some other issues lurking in there, like getting the logic straight for handling the difference between how CLJ and CLJS reference records.

Personally, I'd rather concentrate effort on changing the underlying methods for CLJS compilation so this sort of thing isn't necessary.

Reply:
> Personally, I'd rather concentrate effort on changing the underlying methods for CLJS compilation so this sort of thing isn't necessary.

Sure. That sounds like it could be a separate PR effort too. Just not sure what a good interim solution is rather than including this file in the main jar. Perhaps that's acceptable, or just copy paste this rule definition into the namespaces where it is needed, etc.

Reply:
Just to clarify, I think there are two separate issues here. One has to do with the location of the file, which I suspect is probably solvable with the right project settings or something. I just don't know where the problem lies. The other issue is a deeper one around CLJS compilation and hacks like the use of `require-macros`, where the "solution" is really larger in scope.

I agree that including test_rules_data.clj in the main build is gross, and definitely open to suggestions there.

Reply:
Yes, 

> One has to do with the location of the file, which I suspect is probably solvable with the right project settings or something. I just don't know where the problem lies.

This is something I want to investigate in  #388 . I want to get a better handle on the project structure with regards to cljs builds.

> The other issue is a deeper one around CLJS compilation and hacks like the use of require-macros, where the "solution" is really larger in scope.

I agree. I'm hoping if I can work through #388 I may get a bit more insight since I want to look at these errors we are seeing so far a little closer. 

> I agree that including test_rules_data.clj in the main build is gross, and definitely open to suggestions there.

Not sure, perhaps @WilliamParker has more thoughts.
If we still have `clara.rules.testfacts` in the main jar, I guess this wouldn't be that different. I'd probably at least put a comment on it about it needing to be moved later and only there temporarily for testing reasons.

Reply:
Full stack trace: [Gist](https://gist.github.com/WilliamParker/077b50fbbbca75ccc89a545afee35df0)

I think the key lines here are: 

```

	at clara.rules.compiler$get_cljs_accessors.invokeStatic(compiler.clj:144)
	at clara.rules.compiler$get_cljs_accessors.invoke(compiler.clj:139)
	at clara.rules.compiler$get_fields.invokeStatic(compiler.clj:196)
	at clara.rules.compiler$get_fields.invoke(compiler.clj:189)
	at clara.rules.compiler$field_name__GT_accessors_used.invokeStatic(compiler.clj:325)
	at clara.rules.compiler$field_name__GT_accessors_used.invoke(compiler.clj:321)
```

Basically, we are 

- Passing ClojureScript code to a Clojure macro
- Evaluating that code in a Clojure environment.
- The return value from the previous step contains paths to ClojureScript objects, which we use information from the ClojureScript compiler to reason about.

This analysis is in a way outside the dependency DAG between ClojureScript namespaces, so it wouldn't surprise me if some assumption(s) made by the ClojureScript compiler are being broken.  It also makes me wonder if there is some way to expose this scenario without the main vs test code dichotomy here, which would be a bug.

For an immediate solution, I suspect using rules that don't operate on record types, but rather use primitive data structures, e.g. maps, would work.

Reply:
Just FYI, I think solving this issue is probably best a separate issue. I plan to look over some of the cljs testing stuff with #388 
I've already did some investigation of the problems we are seeing here. I can also cause problems in `master` with compiling these cljs test namespaces with no further changes.

The kicker is just that `lein-cljsbuild` seems to be able to successfully compile these test cljs namespaces, when the `cljs.build.api` and also figwheel cannot. I think something sneaky is just masking the problem in `lein-cljsbuild`. I plan to work more on this and have some more details on #388 within the next few days.